Add automatic evaluation for compositions of inverse hyperbolic functions

ethankward opened the pull request:
#### References to other Issues or PRs
Closes #17421. The more difficult issue with evaluation near branch cuts is not addressed, but there is another issue that covers that (#6137).

#### Brief description of what is fixed or changed
Add automatic evaluation of `atanh(tanh(x)), acosh(cosh(x)), asinh(sinh(x))` for suitable `x`.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* functions
  * atanh(tanh(x)), acosh(cosh(x)) and asinh(sinh(x)) are automatically evaluated
<!-- END RELEASE NOTES -->

eward-sunbelt committed:
add automatic evaluation of atanh(tanh(x)), asinh(sinh(x)), acosh(cosh(x))

oscarbenjamin commented:
Looks good. The diff coverage is apparently 19% though. I'm not sure if I trust the codecov coverage reports in general. Usually the diff coverage seems reasonable but here I would expect a much higher number...

ethankward commented:
Yes, I can see that most of the lines are supposedly red- I can see them being run when I run the tests locally however.

ethankward commented:
Yes, most of the lines are red. I can see they are definitely being run locally when I run the tests.

asmeurer commented:
I would remove the spaces around `*` and `/`. That makes it easier to read the expressions generally.

jksuom commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -926,6 +927,19 @@ def eval(cls, arg):
                     return -cls(-arg)
 
+        if isinstance(arg, sinh):
+            x = arg.args[0]
```
I would first check that `x` is a number and only then go on. Otherwise, I think that automatic simplification should be avoided.

gschintgen commented:
While working on log(exp()) simplification (PR #17301) and similar, I noticed that as_real_imag can be quite eager/aggressive. In particular it unconditionnally expands the expression. Depending on the expression this may be unwanted (computationally or otherwise), especially for automatic evaluation. I settled for a new _match_real_imag() helper in exponential.py. You may find it useful. (Just a suggestion)

It might also be advantageous to not return the imaginary part with I factored out. (See one of the added tests where a further simplification is missed due to this.)

On 16 August 2019 00:02:23 CEST, Ethan Ward <notifications@github.com> wrote:
>#### References to other Issues or PRs
>Closes #17421. The more difficult issue with evaluation near branch
>cuts is not addressed, but there is another issue that covers that
>(#6137).
>
>#### Brief description of what is fixed or changed
>Add automatic evaluation of `atanh(tanh(x)), acosh(cosh(x)),
>asinh(sinh(x))` for suitable `x`.
>
>#### Release Notes
><!-- BEGIN RELEASE NOTES -->
>* functions
>* atanh(tanh(x)), acosh(cosh(x)) and asinh(sinh(x)) are automatically
>evaluated
><!-- END RELEASE NOTES -->
>
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/sympy/sympy/pull/17429
>
>-- Commit Summary --
>
>  * test_issue_17421
>* add automatic evaluation of atanh(tanh(x)), asinh(sinh(x)),
>acosh(cosh(x))
>* test automatic evaluation of atanh(tanh(x)), asinh(sinh(x)),
>acosh(cosh(x))
>
>-- File Changes --
>
>    M sympy/core/tests/test_evalf.py (6)
>    M sympy/functions/elementary/hyperbolic.py (49)
>    M sympy/functions/elementary/tests/test_hyperbolic.py (88)
>
>-- Patch Links --
>
>https://github.com/sympy/sympy/pull/17429.patch
>https://github.com/sympy/sympy/pull/17429.diff
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/sympy/sympy/pull/17429

ethankward commented:
OK. I've modified the code to use `match_real_imag` (and also updated `exponential.py` to remove the initial underscore since it's being used in other files now).

ethankward committed:
Merge branch 'master' into acosh_eval

eward-sunbelt committed:
fix merge problem

gschintgen commented:
Looks good to me.


On 16 August 2019 18:25:31 CEST, Ethan Ward <notifications@github.com> wrote:
>OK. I've modified the code to use `match_real_imag` (and also updated
>`exponential.py` to remove the initial underscore since it's being used
>in other files now).
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/sympy/sympy/pull/17429#issuecomment-522068879

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -926,6 +927,21 @@ def eval(cls, arg):
+                    if f.is_number and f.is_integer:
+                        m = r + I*i - I*pi*f
+                        if f % 2 == 0:
+                            return m
```
We already know that f.is_number because f comes from x and we checked that x.is_number. Likewise if f is floor(...) then f.is_integer *should* always be True (although it might sometimes return None). It is still possible though that f.is_number and f.is_integer but we can't calculate f%2. I would say that the check would be better as:
```python
even = f.is_even
if even is True:
    return m
elif even is False:
    return -m
# Note the 3rd case even is None where we don't evaluate
```

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -926,6 +927,21 @@ def eval(cls, arg):
 
+        if isinstance(arg, sinh):
+            x = arg.args[0]
+            if x.is_number:
```
You can reduce the indentation here by combining the conditions:
```python
if isinstance(arg, sinh) and arg.args[0].is_number:
    ...
```
Or:
```python
if isinstance(arg, sinh):
    x = arg.args[0]
    if not x.is_number:
        return
    ...
```

eward-sunbelt committed:
refactor

oscarbenjamin commented:
Looks good!
